### PR TITLE
[Operator] Add isin op

### PR DESCRIPTION
**Pull request by zhzhcookie**
1. Add isin op.
2. Add unit tests and performance tests of isin op.
3. Optimize the implement of unique op, which is used by isin op.

### CODE REVIEW

File: src/flag_gems/ops/isin.py [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+) -> torch.Tensor:
+    if not torch.is_tensor(in0):
+        assert torch.is_tensor(in1)
+        in0 = torch.tensor(in0, device=in1.device)
+    elif not torch.is_tensor(in1):
+        assert torch.is_tensor(in0)
+        in1 = torch.tensor(in1, device=in0.device)
+    if in0.numel() == 0 or in1.numel() == 0:
+        return torch.zeros_like(in0, dtype=torch.bool)
+    elif in0.numel() <= 12288 and in1.numel() <= 12288:  # 1024 * 12
```
Comment (Bowen12992) [inline src/flag_gems/ops/isin.py:289]: Here why use `1024 * 12` and `1024 * 4096` as the condition ?

File: src/flag_gems/ops/isin.py [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+        return isin_by_comparation(in0, in1, invert)
+    elif assume_unique or in1.numel() <= 4194304:  # 1024 * 4096
```
Comment (Bowen12992) [inline src/flag_gems/ops/isin.py:263]: Here why use `1024 * 12` and `1024 * 4096` as the condition ?

Comment (zhzhcookie) [reply src/flag_gems/ops/isin.py:263]: > Here why use `1024 * 12` and `1024 * 4096` as the condition ?

It's the autotune result.

File: src/flag_gems/__init__.py [hunk 3]
```diff
@@ -51,6 +51,9 @@ def enable(lib=aten_lib):
     lib.impl("gt.Tensor", gt, "CUDA")
     lib.impl("gt.Scalar", gt_scalar, "CUDA")
     lib.impl("isfinite", isfinite, "CUDA")
+    lib.impl("isin.Tensor_Tensor", isin, "CUDA")
```
Comment (Bowen12992) [inline src/flag_gems/__init__.py:54]: This will cause a `double dispatch` but I think it's OK

File: src/flag_gems/__init__.py [hunk 4]
```diff
@@ -51,6 +51,9 @@ def enable(lib=aten_lib):
...
+    lib.impl("isin.Scalar_Tensor", isin, "CUDA")
+    lib.impl("isin.Tensor_Scalar", isin, "CUDA")
```
Comment (Bowen12992) [inline src/flag_gems/__init__.py:59]: This will cause a `double dispatch` but I think it's OK

File: src/flag_gems/ops/isin.py [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+    in1_ravel_ptr: tl.tensor,  # in
+    out_ptr: tl.tensor,  # out
+    M: int,  # num_tasks
+    N: int,  # num_tasks_1
+    BLOCK_M: tl.constexpr,  # tile_size
+    BLOCK_N: tl.constexpr,  # tile_size_1
+    invert: tl.constexpr,
+):
+    row_off = global_pid * BLOCK_M
+    rows = row_off + tl.arange(0, BLOCK_M)[:, None]
```
Comment (Bowen12992) [inline src/flag_gems/ops/isin.py:31]: Can we use slice like `[:，none]` in a triton.jit kernel now?

File: src/flag_gems/ops/isin.py [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+    N: int,  # num_tasks_1
+    BLOCK_M: tl.constexpr,  # tile_size
+    BLOCK_N: tl.constexpr,  # tile_size_1
+    tiles_per_cta: int,
+    one_tile_per_cta: tl.constexpr,
+    invert: tl.constexpr,
+):
+    pid = tl.program_id(0)
+    num_ctas = tl.num_programs(0)
+    if one_tile_per_cta:  # monolitic kernel style
```
Comment (Bowen12992) [inline src/flag_gems/ops/isin.py]: Why can't we also use `grid-stride-loop style kernel` in this case

File: src/flag_gems/ops/isin.py [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+    M: int,  # num_tasks
+    N: int,  # num_tasks_1
+    BLOCK_M: tl.constexpr,  # tile_size
+    tiles_per_cta: int,
+    one_tile_per_cta: tl.constexpr,
+    invert: tl.constexpr,
+):
+    pid = tl.program_id(0)
+    num_ctas = tl.num_programs(0)
+    if one_tile_per_cta:  # monolitic kernel style
```
Comment (Bowen12992) [inline src/flag_gems/ops/isin.py]: dito

File: src/flag_gems/ops/isin.py [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+
+
+@triton.jit
+def all_fn(a, b):
+    return a and b
+
+
+@triton.jit
+def any_fn(a, b):
+    return a or b
```
Comment (Bowen12992) [inline src/flag_gems/ops/isin.py]: Maybe we can
```suggestion
from .all import reduce_all as all_fn
from .any import reduce_any as any_fn
```

Comment (zhzhcookie) [reply src/flag_gems/ops/isin.py]: > Maybe we can

Done

File: src/flag_gems/ops/isin.py (hunk 6 above)
Comment (zhzhcookie) [reply src/flag_gems/ops/isin.py]: > Why can't we also use `grid-stride-loop style kernel` in this case

Done

File: src/flag_gems/ops/isin.py (hunk 5 above)
Comment (zhzhcookie) [reply src/flag_gems/ops/isin.py:31]: > Can we use slice like `[:，none]` in a triton.jit kernel now?

No, we can only use this to broadcast, but not slice.

**Review by Bowen12992 — DISMISSED** [review]
LGTM,  just wait for `libentry` bug fixed and pull the master :)

File: src/flag_gems/ops/isin.py [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+        BLOCK_M, BLOCK_N, num_warps = launch_arg(1, 256, N, 4)
+    elif M <= 3072:
+        BLOCK_M, BLOCK_N, num_warps = launch_arg(2, 256, N, 4)
+    elif M <= 6144:
+        BLOCK_M, BLOCK_N, num_warps = launch_arg(4, 128, N, 4)
+    elif M <= 9216:
+        BLOCK_M, BLOCK_N, num_warps = launch_arg(4, 256, N, 8)
+    else:
+        BLOCK_M, BLOCK_N, num_warps = launch_arg(4, 128, N, 4)
+    num_ctas = min(65536, triton.cdiv(M, BLOCK_M))
```
Comment (tongxin) [inline src/flag_gems/ops/isin.py]: If this number is to be used as grid.y, then the current limit is 65535.

File: src/flag_gems/ops/isin.py [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+    in0: torch.tensor,
+    in1: torch.tensor,
+    invert: bool,
+):
+    in0_ravel = in0.contiguous().ravel()
+    in1_ravel = in1.contiguous().ravel()
+    M = in0.numel()
+    N = in1.numel()
+    if M <= 1024:
+        BLOCK_M, BLOCK_N, num_warps = launch_arg(1, 256, N, 4)
```
Comment (tongxin) [inline src/flag_gems/ops/isin.py:94]: The reduction dim's tile size BLOCK_N is considerably small. Triton has pretty competitive reduction builtin. Is there room for larger reduction tiles?

File: src/flag_gems/ops/isin.py [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+    if unique_in1:
+        in1_ravel, _, _ = _unique2(
+            in1, sorted=True, return_inverse=False, return_counts=False
+        )
+    else:
+        in1_ravel, _ = torch.sort(in1.ravel())
+    # launch kernel func
+    M = in0_ravel.numel()
+    N = in1_ravel.numel()
+    if M <= 1048576:  # 2 ** 20 = 1024 * 1024
```
Comment (tongxin) [inline src/flag_gems/ops/isin.py:213]: I don't understand why BLOCK_N is dependent on M.

File: src/flag_gems/ops/isin.py [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+    i0 = global_pid * BLOCK_M + r
+    mask = i0 < M
+
+    # load in0_ravel
+    in0_ravel = tl.load(in0_ravel_ptr + i0, mask=mask)
+
+    # binary search: lower_bound
+    out = tl.zeros_like(r).to(tl.int1)
+    start = tl.zeros_like(r)
+    end = start + N
```
Comment (tongxin) [inline src/flag_gems/ops/isin.py:145]: My experience is end should start with start + N - 1

File: src/flag_gems/ops/isin.py [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+    while_mask = start < end
+    while tl.sum(tl.where(while_mask, out, 1)) != BLOCK_M:
```
Comment (tongxin) [inline src/flag_gems/ops/isin.py]: To save an extra reduction at the end of each iteration, I would rather use a fixed loop bound log2(N) + 1 and make sure that start and end meet in the end.

File: src/flag_gems/ops/isin.py (hunk 12 above)
Comment (tongxin) [reply src/flag_gems/ops/isin.py:145]: The code is correct. Ignore this line;)

File: src/flag_gems/ops/isin.py [hunk 14]
```diff
@@ -0,0 +1,259 @@
...
+    out_ptr: tl.tensor,  # out
+    M: int,  # num_tasks
+    N: int,  # num_tasks_1
+    BLOCK_M: tl.constexpr,  # tile_size
+    BLOCK_N: tl.constexpr,  # tile_size_1
+    tiles_per_cta: int,
+    invert: tl.constexpr,
+):
+    pid = tl.program_id(0)
+    num_ctas = tl.num_programs(0)
```
Comment (tongxin) [inline src/flag_gems/ops/isin.py]: num_ctas is a Triton reserved key word. We'd better rename it.

Comment (zhzhcookie) [reply src/flag_gems/ops/isin.py]: > num_ctas is a Triton reserved key word. We'd better rename it.

done

File: src/flag_gems/ops/isin.py (hunk 13 above)
Comment (zhzhcookie) [reply src/flag_gems/ops/isin.py]: > To save an extra reduction at the end of each iteration, I would rather use a fixed loop bound log2(N) + 1 and make sure that start and end meet in the end.

done

File: src/flag_gems/ops/isin.py (hunk 12 above)
Comment (zhzhcookie) [reply src/flag_gems/ops/isin.py:145]: > My experience is end should start with start + N - 1

[start, start + N) is also ok.

File: src/flag_gems/ops/isin.py (hunk 10 above)
Comment (zhzhcookie) [reply src/flag_gems/ops/isin.py:94]: > The reduction dim's tile size BLOCK_N is considerably small. Triton has pretty competitive reduction builtin. Is there room for larger reduction tiles?

It is the autotune result.

**Review by Bowen12992 — APPROVED** [review]
